python/lsst/metaserv/metaREST_v0.py

```diff
@@ -0,0 +1,176 @@
+            else: # default format is application/json
+                retStr += "%s:%s " % (cursor.description[x][0], row[x])
+        if fmt == "application/json":
+            retStr = json.dumps(retStr)
+    cursor.close()
+    return retStr
+
+def runDbQueryM(query, optTuple=None, notFoundMsg='Not found'):
+    '''Runs query that returns many rows.'''
+    cursor = con.cursor()
+    if optTuple:
+        cursor.execute(query, optTuple)
+    else:
+        cursor.execute(query)
+    rows = cursor.fetchall()
+    fmt = request.accept_mimetypes.best_match(['application/json', 'text/html'])
+    if len(rows) == 0:
+        retStr = notFoundMsg
+    else:
+        if fmt == 'text/html':
+            retStr = "<br />".join(str(r[0]) for r in rows)
+        else: # default format is application/json
+            ret = " ".join(str(r[0]) for r in rows)
+            retStr = json.dumps(ret)
+    cursor.close()
+    return retStr
+
+@metaREST.route('/', methods=['GET'])
+def getRoot():
+    fmt = request.accept_mimetypes.best_match(['application/json', 'text/html'])
+    if fmt == 'text/html':
+        return ("LSST Metadata Service v0 here. I currently support: "
+                "<a href='db'>/db</a> and <a href='image'>/image</a>.")
+    return "LSST Metadata Service v0 here. I currently support: /db and /image."
+
+@metaREST.route('/db', methods=['GET'])
+def getDb():
+    '''Lists types of databases (that have at least one database).'''
+    # todo: this will currently fail if Repo table does not exist
+    return runDbQueryM(
+        "SELECT DISTINCT lsstLevel FROM Repo WHERE repoType = 'db'",
+        None,
+        "No types with at least one database found.")
+
+@metaREST.route('/db/<string:lsstLevel>', methods=['GET'])
+def getDbPerType(lsstLevel):
+    '''Lists databases for a given type.'''
+    # todo: this will currently fail if Repo table does not exist
+    return runDbQueryM(
+        "SELECT dbName FROM Repo JOIN DbMeta on (repoId=dbMetaId) "
+        "WHERE lsstLevel = %s",
+        (lsstLevel,),
+        "No database found.")
+
+@metaREST.route('/db/<string:lsstLevel>/<string:dbName>', methods=['GET'])
+def getDbPerTypeDbName(lsstLevel, dbName):
+    '''Retrieves information about one database.'''
+    # We don't use lsstLevel here because db names are unique across all types.
+    return runDbQuery1(
+        "SELECT Repo.*, DbMeta.* "
+        "FROM Repo JOIN DbMeta on (repoId=dbMetaId) "
+        "WHERE dbName = %s",
+        (dbName,),
+        "Database '%s' not found." % dbName)
+
+@metaREST.route('/db/<string:lsstLevel>/<string:dbName>/tables', methods=['GET'])
+def getDbPerTypeDbNameTables(lsstLevel, dbName):
+    '''Lists table names in a given database.'''
+    return runDbQueryM(
+        "SELECT table_name FROM information_schema.tables "
+        "WHERE table_schema= %s ",
+        (dbName,),
+        "No tables found in database '%s'." % dbName)
+
+@metaREST.route('/db/<string:lsstLevel>/<string:dbName>/tables/' +
```

Comment:
Plus is not really needed at the end of the line, string literals will be concatenated without it.